sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_system_client_test.cpp

```diff
@@ -821,4 +821,145 @@ namespace Azure { namespace Storage { namespace Test {
+    deletedFileClient.Create();
+    deletedFileClient.Delete();
+    nonDeletedFileClient.Create();
+    deletedDirectoryClient.Create();
+    deletedDirectoryClient.DeleteEmpty();
+    nonDeletedDirectoryClient.Create();
+
+    {
+      auto paths = m_fileSystemClient->ListDeletedPaths().DeletedPaths;
+      EXPECT_EQ(2, paths.size());
+      EXPECT_EQ(deletedDirectoryName, paths[0].Name);
+      EXPECT_TRUE(paths[0].DeletionId.HasValue());
+      EXPECT_TRUE(paths[0].DeletedOn.HasValue());
+      EXPECT_TRUE(paths[0].RemainingRetentionDays.HasValue());
+      EXPECT_EQ(deletedFilename, paths[1].Name);
+      EXPECT_TRUE(paths[1].DeletionId.HasValue());
+      EXPECT_TRUE(paths[1].DeletedOn.HasValue());
+      EXPECT_TRUE(paths[1].RemainingRetentionDays.HasValue());
+    }
+    //  List max result
+    {
+      Files::DataLake::ListDeletedPathsOptions options;
+      options.PageSizeHint = 1;
+      std::vector<Files::DataLake::Models::PathDeletedItem> paths;
+      for (auto pageResult = m_fileSystemClient->ListDeletedPaths(options); pageResult.HasPage();
+           pageResult.MoveToNextPage())
+      {
+        paths.insert(paths.end(), pageResult.DeletedPaths.begin(), pageResult.DeletedPaths.end());
+        EXPECT_EQ(1, pageResult.DeletedPaths.size());
+      }
+      EXPECT_EQ(2, paths.size());
+    }
+    // prefix works
+    {
+      const std::string directoryName = GetTestNameLowerCase() + "_prefix";
+      const std::string filename = "file";
+
+      auto directoryClient = m_fileSystemClient->GetDirectoryClient(directoryName);
+      directoryClient.Create();
+      auto fileClient = directoryClient.GetFileClient(filename);
+      fileClient.Create();
+      fileClient.Delete();
+
+      Files::DataLake::ListDeletedPathsOptions options;
+      options.PathPrefix = directoryName;
+      auto paths = m_fileSystemClient->ListDeletedPaths(options).DeletedPaths;
+      EXPECT_EQ(1, paths.size());
+      EXPECT_EQ(directoryName + "/" + filename, paths[0].Name);
+      EXPECT_TRUE(paths[0].DeletionId.HasValue());
+      EXPECT_TRUE(paths[0].DeletedOn.HasValue());
+      EXPECT_TRUE(paths[0].RemainingRetentionDays.HasValue());
+    }
+  }
+
+  TEST_F(DataLakeFileSystemClientTest, Undelete)
+  {
+    const std::string directoryName = GetTestNameLowerCase() + "_dir";
+    const std::string subFileName = "sub_file";
+
+    auto directoryClient = m_fileSystemClient->GetDirectoryClient(directoryName);
+    directoryClient.Create();
+    auto subFileClient = m_fileSystemClient->GetFileClient(directoryName + "/" + subFileName);
+    subFileClient.Create();
+    std::shared_ptr<Files::DataLake::Models::PathProperties> properties
+        = std::make_shared<Files::DataLake::Models::PathProperties>(
+            directoryClient.GetProperties().Value);
+    std::shared_ptr<Files::DataLake::Models::PathProperties> subFileProperties
+        = std::make_shared<Files::DataLake::Models::PathProperties>(
+            subFileClient.GetProperties().Value);
```

Comment:
Maybe it doesn't matter. But why do you use shared pointers here anyway?

Reply:
If don't use shared pointers, It may cause big stack usage and can't pass the ci test.